File: src/ai_agent_sdk/transport.py
from __future__ import annotations

import logging
from typing import Any

import httpx

from ai_agent_sdk.config import SDKConfig
from ai_agent_sdk.exceptions import TransportError, raise_for_status
from ai_agent_sdk.models import RateLimitInfo

logger = logging.getLogger(__name__)


class Transport:
    def __init__(self, config: SDKConfig) -> None:
        self._config = config
        self._client: httpx.Client | None = None
        self._last_rate_limit: RateLimitInfo | None = None

    @property
    def client(self) -> httpx.Client:
        if self._client is None or self._client.is_closed:
            headers = dict(self._config.custom_headers)
            if self._config.api_key:
                headers["Authorization"] = f"Bearer {self._config.api_key}"
            self._client = httpx.Client(
                base_url=self._config.base_url,
                headers=headers,
                timeout=httpx.Timeout(self._config.timeout),
            )
        return self._client

    @property
    def last_rate_limit(self) -> RateLimitInfo | None:
        return self._last_rate_limit

    def _parse_rate_limit(self, response: httpx.Response) -> RateLimitInfo | None:
        limit = response.headers.get("X-RateLimit-Limit")
        remaining = response.headers.get("X-RateLimit-Remaining")
        if limit is not None and remaining is not None:
            info = RateLimitInfo(limit=int(limit), remaining=int(remaining))
            self._last_rate_limit = info
            return info
        return None

    def request(
        self,
        method: str,
        path: str,
        *,
        json: Any | None = None,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        try:
            if self._config.log_requests:
                logger.info("Request: %s %s params=%s", method, path, params)
            response = self.client.request(
                method,
                path,
                json=json,
                params=params,
                headers=headers,
            )
            self._parse_rate_limit(response)
            if self._config.log_responses:
                logger.info("Response: %s %s", response.status_code, response.text[:500])
            raise_for_status(response.status_code, response.text)
            return response
        except httpx.HTTPError as exc:
            raise TransportError(str(exc)) from exc

    def get(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        return self.request("GET", path, params=params, headers=headers)

    def post(
        self,
        path: str,
        *,
        json: Any | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        return self.request("POST", path, json=json, headers=headers)

    def put(
        self,
        path: str,
        *,
        json: Any | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        return self.request("PUT", path, json=json, headers=headers)

    def delete(
        self,
        path: str,
        *,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        return self.request("DELETE", path, headers=headers)

    def close(self) -> None:
        if self._client is not None and not self._client.is_closed:
            self._client.close()
            self._client = None


class AsyncTransport:
    def __init__(self, config: SDKConfig) -> None:
        self._config = config
        self._client: httpx.AsyncClient | None = None
        self._last_rate_limit: RateLimitInfo | None = None

    @property
    def client(self) -> httpx.AsyncClient:
        if self._client is None or self._client.is_closed:
            headers = dict(self._config.custom_headers)
            if self._config.api_key:
                headers["Authorization"] = f"Bearer {self._config.api_key}"
            self._client = httpx.AsyncClient(
                base_url=self._config.base_url,
                headers=headers,
                timeout=httpx.Timeout(self._config.timeout),
            )
        return self._client

    @property
    def last_rate_limit(self) -> RateLimitInfo | None:
        return self._last_rate_limit

    def _parse_rate_limit(self, response: httpx.Response) -> RateLimitInfo | None:
        limit = response.headers.get("X-RateLimit-Limit")
        remaining = response.headers.get("X-RateLimit-Remaining")
        if limit is not None and remaining is not None:
            info = RateLimitInfo(limit=int(limit), remaining=int(remaining))
            self._last_rate_limit = info
            return info
        return None

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: Any | None = None,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        try:
            if self._config.log_requests:
                logger.info("Request: %s %s params=%s", method, path, params)
            response = await self.client.request(
                method,
                path,
                json=json,
                params=params,
                headers=headers,
            )
            self._parse_rate_limit(response)
            if self._config.log_responses:
                logger.info("Response: %s %s", response.status_code, response.text[:500])
            raise_for_status(response.status_code, response.text)
            return response
        except httpx.HTTPError as exc:
            raise TransportError(str(exc)) from exc

    async def get(
        self,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        return await self.request("GET", path, params=params, headers=headers)

    async def post(
        self,
        path: str,
        *,
        json: Any | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        return await self.request("POST", path, json=json, headers=headers)

    async def put(
        self,
        path: str,
        *,
        json: Any | None = None,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        return await self.request("PUT", path, json=json, headers=headers)

    async def delete(
        self,
        path: str,
        *,
        headers: dict[str, str] | None = None,
    ) -> httpx.Response:
        return await self.request("DELETE", path, headers=headers)

    async def close(self) -> None:
        if self._client is not None and not self._client.is_closed:
            await self._client.aclose()
            self._client = None

File: rust-sdk/src/transport.rs
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::Duration;

use reqwest::header::{HeaderMap, HeaderName, HeaderValue};
use reqwest::{Client, Method, Response};
use tracing::debug;

use crate::config::SDKConfig;
use crate::errors::{raise_for_status, AgentSDKError, Result};
use crate::models::RateLimitInfo;

pub struct Transport {
    client: Client,
    config: SDKConfig,
    last_rate_limit: Mutex<Option<RateLimitInfo>>,
}

impl Transport {
    pub fn new(config: &SDKConfig) -> Result<Self> {
        let mut default_headers = HeaderMap::new();
        if !config.api_key.is_empty() {
            let val = format!("Bearer {}", config.api_key);
            default_headers.insert(
                reqwest::header::AUTHORIZATION,
                HeaderValue::from_str(&val).map_err(|e| {
                    AgentSDKError::Configuration(format!("Invalid API key header: {e}"))
                })?,
            );
        }
        for (k, v) in &config.custom_headers {
            let name = HeaderName::from_bytes(k.as_bytes())
                .map_err(|e| AgentSDKError::Configuration(format!("Invalid header name: {e}")))?;
            let value = HeaderValue::from_str(v)
                .map_err(|e| AgentSDKError::Configuration(format!("Invalid header value: {e}")))?;
            default_headers.insert(name, value);
        }

        let client = Client::builder()
            .timeout(Duration::from_secs(config.timeout_secs))
            .default_headers(default_headers)
            .build()
            .map_err(|e| AgentSDKError::Transport(format!("Failed to build HTTP client: {e}")))?;

        Ok(Self {
            client,
            config: config.clone(),
            last_rate_limit: Mutex::new(None),
        })
    }

    fn parse_rate_limit(&self, response: &Response) -> Option<RateLimitInfo> {
        let limit = response
            .headers()
            .get("X-RateLimit-Limit")
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.parse::<u64>().ok());
        let remaining = response
            .headers()
            .get("X-RateLimit-Remaining")
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.parse::<u64>().ok());

        if let (Some(limit), Some(remaining)) = (limit, remaining) {
            let info = RateLimitInfo {
                limit,
                remaining,
                reset_at: None,
            };
            if let Ok(mut guard) = self.last_rate_limit.lock() {
                *guard = Some(info.clone());
            }
            Some(info)
        } else {
            None
        }
    }

    pub fn last_rate_limit(&self) -> Option<RateLimitInfo> {
        self.last_rate_limit
            .lock()
            .ok()
            .and_then(|guard| guard.clone())
    }

    pub async fn request(
        &self,
        method: Method,
        path: &str,
        json_body: Option<&serde_json::Value>,
        params: Option<&HashMap<String, String>>,
        extra_headers: Option<&HashMap<String, String>>,
    ) -> Result<Response> {
        let url = format!("{}{}", self.config.base_url.trim_end_matches('/'), path);

        if self.config.log_requests {
            debug!("-> {} {} params={:?}", method, url, params);
        }

        let mut req = self.client.request(method.clone(), &url);

        if let Some(p) = params {
            req = req.query(p);
        }
        if let Some(body) = json_body {
            req = req.json(body);
        }
        if let Some(headers) = extra_headers {
            for (k, v) in headers {
                req = req.header(k.as_str(), v.as_str());
            }
        }

        let response = req.send().await.map_err(|e| {
            AgentSDKError::Transport(format!("Request failed: {e}"))
        })?;

        self.parse_rate_limit(&response);

        let status = response.status().as_u16();

        if self.config.log_responses {
            debug!("<- {} [{}]", url, status);
        }

        if status >= 400 {
            let body_text = response.text().await.unwrap_or_default();
            raise_for_status(status, &body_text)?;
            unreachable!();
        }

        Ok(response)
    }

    pub async fn get(
        &self,
        path: &str,
        params: Option<&HashMap<String, String>>,
        headers: Option<&HashMap<String, String>>,
    ) -> Result<Response> {
        self.request(Method::GET, path, None, params, headers).await
    }

    pub async fn post(
        &self,
        path: &str,
        json_body: Option<&serde_json::Value>,
        headers: Option<&HashMap<String, String>>,
    ) -> Result<Response> {
        self.request(Method::POST, path, json_body, None, headers)
            .await
    }

    pub async fn put(
        &self,
        path: &str,
        json_body: Option<&serde_json::Value>,
        headers: Option<&HashMap<String, String>>,
    ) -> Result<Response> {
        self.request(Method::PUT, path, json_body, None, headers)
            .await
    }

    pub async fn delete(
        &self,
        path: &str,
        headers: Option<&HashMap<String, String>>,
    ) -> Result<Response> {
        self.request(Method::DELETE, path, None, None, headers)
            .await
    }
}
